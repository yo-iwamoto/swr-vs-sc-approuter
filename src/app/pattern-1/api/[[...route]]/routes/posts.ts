import { prisma } from "@/lib/prisma";
import { zValidator } from "@hono/zod-validator";
import { Hono } from "hono";
import { z } from "zod";
import { authMiddleware, currentUserId } from "../lib/auth";

const app = new Hono();

app.use("*", authMiddleware);

export const postsRoute = app
  .get("/", async (c) => {
    const posts = await prisma.post.findMany({
      include: { User: true },
      orderBy: { createdAt: "desc" },
    });

    return c.json({ posts });
  })
  .post(
    "/",
    zValidator(
      "json",
      z.object({
        content: z.string().min(1).max(140),
      }),
    ),
    async (c) => {
      const { content } = c.req.valid("json");
      const userId = await currentUserId(c);
      const post = await prisma.post.create({
        data: { content, userId },
      });
      return c.json({ post }, 201);
    },
  )
  .delete("/:id", async (c) => {
    const id = c.req.param("id");
    const userId = await currentUserId(c);
    await prisma.post.delete({ where: { id, userId } });
    return c.json({ status: "ok" });
  })
  .post("/:id/likes", async (c) => {
    const id = c.req.param("id");
    const userId = await currentUserId(c);
    const like = await prisma.like.create({
      data: { userId, postId: id },
    });
    return c.json({ like }, 201);
  })
  .delete("/:id/likes", async (c) => {
    const id = c.req.param("id");
    const userId = await currentUserId(c);
    await prisma.like.deleteMany({ where: { postId: id, userId } });
    return c.json({ status: "ok" });
  })
  .get("/timeline", async (c) => {
    const userId = await currentUserId(c);
    const followingUserIds = (
      await prisma.follow.findMany({
        where: { followingId: userId },
        select: { followerId: true },
      })
    ).map((follow) => follow.followerId);
    const posts = await prisma.post.findMany({
      where: { userId: { in: [...followingUserIds, userId] } },
      include: { User: true },
      orderBy: { createdAt: "desc" },
    });
    return c.json({ posts, followingUserIds });
  });
