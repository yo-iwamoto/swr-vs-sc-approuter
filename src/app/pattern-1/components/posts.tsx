"use client";

import type { PostApiResponse } from "@/lib/api";
import {
  Base,
  Button,
  Cluster,
  Dropdown,
  DropdownContent,
  DropdownTrigger,
  FaCaretDownIcon,
  FaUserIcon,
  Loader,
} from "smarthr-ui";
import { mutate } from "swr";
import { useMeQuery } from "../queries/use-me-query";
import { usePostsQuery } from "../queries/use-posts-query";
import { useDeletePostMutation } from "../mutations/use-delete-post-mutation";

export function Posts() {
  const postsQuery = usePostsQuery();

  if (postsQuery.error !== undefined) {
    return <p>Error</p>;
  }

  if (postsQuery.data === undefined) {
    return (
      <div className="h-80 grid place-items-center">
        <Loader />
      </div>
    );
  }

  if (postsQuery.data.posts.length === 0) {
    return <p className="text-center">ポストがありません</p>;
  }

  return (
    <ul className="grid gap-3 max-w-2xl mx-auto w-full">
      {postsQuery.data.posts.map((post) => (
        <Post key={post.id} post={post} />
      ))}
    </ul>
  );
}

type PostProps = {
  post: PostApiResponse;
};

function Post({ post }: PostProps) {
  const meQuery = useMeQuery();

  const mutation = useDeletePostMutation();

  const onClickDelete = async () => {
    if (!confirm("本当に削除しますか？")) return;

    await mutation.trigger({ id: post.id });
    await mutate(["posts"]);
  };

  if (meQuery.data === undefined) return false;

  const isMyPost = meQuery.data.user.id === post.userId;

  return (
    <Base as="li" className="p-4 grid gap-2">
      <Cluster align="center" justify="space-between">
        <p className="font-bold">
          <Cluster align="center">
            <FaUserIcon />
            <span>{post.User.username}</span>
          </Cluster>
        </p>

        {isMyPost && (
          <Dropdown>
            <DropdownTrigger>
              <Button size="s" type="button" suffix={<FaCaretDownIcon />}>
                操作
              </Button>
            </DropdownTrigger>
            <DropdownContent>
              <ul>
                <li>
                  <Button type="button" onClick={onClickDelete}>
                    削除
                  </Button>
                </li>
              </ul>
            </DropdownContent>
          </Dropdown>
        )}
      </Cluster>

      <p>{post.content}</p>
    </Base>
  );
}
