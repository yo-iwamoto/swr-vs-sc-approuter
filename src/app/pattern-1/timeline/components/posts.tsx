"use client";

import { Post } from "@/app/pattern-1/components/post";
import { Loader } from "smarthr-ui";
import { useTimelineQuery } from "../queries/use-timeline-query";

export function Posts() {
  const timelineQuery = useTimelineQuery();

  if (timelineQuery.error !== undefined) {
    return <p>Error</p>;
  }

  if (timelineQuery.data === undefined) {
    return (
      <div className="h-80 grid place-items-center">
        <Loader />
      </div>
    );
  }

  if (timelineQuery.data.posts.length === 0) {
    const message =
      timelineQuery.data.followingUserIds.length === 0
        ? "まだ誰もフォローしていません"
        : "ポストがありません";

    return <p className="text-center">{message}</p>;
  }

  return (
    <ul className="grid gap-3 max-w-2xl mx-auto w-full">
      {timelineQuery.data.posts.map((post) => (
        <Post key={post.id} post={post} />
      ))}
    </ul>
  );
}
