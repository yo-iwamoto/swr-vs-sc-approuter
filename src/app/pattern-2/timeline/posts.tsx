import { type PostApiResponse, api } from "@/lib/api";
import { Post } from "../components/post";
import { callApi } from "../server/call-api";

export async function Posts() {
  const { posts }: { posts: PostApiResponse[] } = await callApi(
    api.posts.timeline.$url().toString(),
  );

  if (posts.length === 0) {
    return <p className="text-center">ポストがありません</p>;
  }

  return (
    <ul className="grid gap-3 max-w-2xl mx-auto w-full">
      {posts.map((post) => (
        <Post key={post.id} post={post} />
      ))}
    </ul>
  );
}
