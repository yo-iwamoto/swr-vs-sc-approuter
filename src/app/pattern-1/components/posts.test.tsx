import { render, screen } from "@testing-library/react";
import { Posts } from "./posts";

const useMeQueryMock = vi.hoisted(() =>
  vi.fn(() => ({ data: { user: { id: "1" } } })),
);
vi.mock("@/app/pattern-1/queries/use-me-query", () => ({
  useMeQuery: useMeQueryMock,
}));

const usePostsQueryMock = vi.hoisted(() => vi.fn());
vi.mock("@/app/pattern-1/queries/use-posts-query", () => ({
  usePostsQuery: usePostsQueryMock,
}));

const triggerDeletePostMock = vi.fn();
vi.mock("@/app/pattern-1/mutations/use-delete-post-mutation", () => ({
  useDeletePostMutation: vi.fn(() => ({
    trigger: triggerDeletePostMock,
  })),
}));

const mutateMock = vi.hoisted(() => vi.fn());
vi.mock("swr", () => ({ mutate: mutateMock }));

describe("Posts", () => {
  it("ポストが1件もない時、ポストがありませんと表示されること", () => {
    usePostsQueryMock.mockReturnValue({ data: { posts: [] } });

    render(<Posts />);

    expect(screen.getByText("ポストがありません")).toBeDefined();
  });

  it("ポストが取得中の時、ローダーが表示されること", () => {
    usePostsQueryMock.mockReturnValue({ data: undefined });

    render(<Posts />);

    expect(screen.getByRole("status")).toHaveTextContent("処理中");
  });

  it("ポストが取得された時、ポストが全件表示されること", () => {
    usePostsQueryMock.mockReturnValue({
      data: {
        posts: [
          { id: "1", User: { username: "user-1" }, content: "Hello world" },
          { id: "2", User: { username: "user-2" }, content: "Second post" },
        ],
      },
    });

    render(<Posts />);

    expect(screen.getAllByRole("listitem")).toHaveLength(2);
    expect(screen.getAllByRole("listitem")[0]).toHaveTextContent(/user-1/);
    expect(screen.getAllByRole("listitem")[0]).toHaveTextContent(/Hello world/);
  });
});
