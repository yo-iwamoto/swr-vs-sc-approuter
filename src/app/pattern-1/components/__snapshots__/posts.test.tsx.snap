// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Posts > snapshot 1`] = `
<DocumentFragment>
  <ul
    class="grid gap-3 max-w-2xl mx-auto w-full"
  >
    <li
      class="smarthr-ui-Base shr-bg-white forced-colors:shr-border-shorthand contrast-more:shr-border-high-contrast shr-rounded-l shr-shadow-layer-1 p-4 grid gap-2"
    >
      <div
        class="shr-flex-wrap [&:empty]:shr-gap-0 shr-flex shr-gap-y-0.5 shr-gap-x-0.5 shr-items-center shr-justify-between"
      >
        <div
          class="flex items-center gap-2"
        >
          <a
            class="font-bold hover:underline"
            href="/pattern-1/user/undefined"
          >
            <span
              class="shr-flex-wrap [&:empty]:shr-gap-0 shr-flex shr-gap-y-0.5 shr-gap-x-0.5 shr-items-center"
            >
              <svg
                aria-hidden="true"
                class="smarthr-ui-Icon group-[]/iconWrapper:shr-shrink-0 group-[]/iconWrapper:shr-translate-y-[0.125em] forced-colors:shr-fill-[CanvasText]"
                fill="currentColor"
                focusable="false"
                height="1em"
                role="img"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 448 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M224 256A128 128 0 1 0 224 0a128 128 0 1 0 0 256zm-45.7 48C79.8 304 0 383.8 0 482.3C0 498.7 13.3 512 29.7 512H418.3c16.4 0 29.7-13.3 29.7-29.7C448 383.8 368.2 304 269.7 304H178.3z"
                />
              </svg>
              <span>
                user-1
              </span>
            </span>
          </a>
          <span
            class="text-sm text-gray-600"
          >
            NaN-NaN-NaN NaN:NaN
          </span>
          <span
            class="text-sm flex items-center gap-1"
          >
            <svg
              aria-hidden="true"
              class="smarthr-ui-Icon group-[]/iconWrapper:shr-shrink-0 group-[]/iconWrapper:shr-translate-y-[0.125em] forced-colors:shr-fill-[CanvasText] text-yellow-500"
              fill="currentColor"
              focusable="false"
              height="1em"
              role="img"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 576 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M316.9 18C311.6 7 300.4 0 288.1 0s-23.4 7-28.8 18L195 150.3 51.4 171.5c-12 1.8-22 10.2-25.7 21.7s-.7 24.2 7.9 32.7L137.8 329 113.2 474.7c-2 12 3 24.2 12.9 31.3s23 8 33.8 2.3l128.3-68.5 128.3 68.5c10.8 5.7 23.9 4.9 33.8-2.3s14.9-19.3 12.9-31.3L438.5 329 542.7 225.9c8.6-8.5 11.7-21.2 7.9-32.7s-13.7-19.9-25.7-21.7L381.2 150.3 316.9 18z"
              />
            </svg>
            <span />
          </span>
        </div>
        <button
          class="disabled:shr-cursor-not-allowed disabled:shr-bg-clip-padding [&_.smarthr-ui-Icon]:forced-colors:disabled:shr-fill-[GrayText] shr-box-border shr-cursor-pointer shr-inline-flex shr-justify-center shr-items-center shr-gap-0.5 shr-text-center shr-whitespace-nowrap shr-rounded-m shr-border-shorthand shr-font-inherit shr-font-bold shr-leading-none focus-visible:shr-focus-indicator contrast-more:shr-border-high-contrast [&_.smarthr-ui-Icon]:shr-block [&_svg]:shr-block shr-p-0.5 shr-text-sm shr-min-h-[calc(theme(fontSize.sm)+theme(spacing.1)+theme(borderWidth.2))] shr-border-main shr-bg-main shr-text-white focus-visible:shr-border-main-darken focus-visible:shr-bg-main-darken hover:shr-border-main-darken hover:shr-bg-main-darken disabled:shr-border-main/50 disabled:shr-bg-main/50 disabled:shr-text-white/50 smarthr-ui-Button"
          type="button"
        >
          <span
            class="shr-min-w-0 shr-leading-[0]"
          >
            <div
              class="flex items-center gap-2"
            >
              <svg
                aria-hidden="true"
                class="smarthr-ui-Icon group-[]/iconWrapper:shr-shrink-0 group-[]/iconWrapper:shr-translate-y-[0.125em] forced-colors:shr-fill-[CanvasText]"
                fill="currentColor"
                focusable="false"
                height="1em"
                role="img"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 576 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M316.9 18C311.6 7 300.4 0 288.1 0s-23.4 7-28.8 18L195 150.3 51.4 171.5c-12 1.8-22 10.2-25.7 21.7s-.7 24.2 7.9 32.7L137.8 329 113.2 474.7c-2 12 3 24.2 12.9 31.3s23 8 33.8 2.3l128.3-68.5 128.3 68.5c10.8 5.7 23.9 4.9 33.8-2.3s14.9-19.3 12.9-31.3L438.5 329 542.7 225.9c8.6-8.5 11.7-21.2 7.9-32.7s-13.7-19.9-25.7-21.7L381.2 150.3 316.9 18z"
                />
              </svg>
              <span>
                いいね
              </span>
            </div>
          </span>
        </button>
      </div>
      <p>
        Hello world
      </p>
    </li>
    <li
      class="smarthr-ui-Base shr-bg-white forced-colors:shr-border-shorthand contrast-more:shr-border-high-contrast shr-rounded-l shr-shadow-layer-1 p-4 grid gap-2"
    >
      <div
        class="shr-flex-wrap [&:empty]:shr-gap-0 shr-flex shr-gap-y-0.5 shr-gap-x-0.5 shr-items-center shr-justify-between"
      >
        <div
          class="flex items-center gap-2"
        >
          <a
            class="font-bold hover:underline"
            href="/pattern-1/user/undefined"
          >
            <span
              class="shr-flex-wrap [&:empty]:shr-gap-0 shr-flex shr-gap-y-0.5 shr-gap-x-0.5 shr-items-center"
            >
              <svg
                aria-hidden="true"
                class="smarthr-ui-Icon group-[]/iconWrapper:shr-shrink-0 group-[]/iconWrapper:shr-translate-y-[0.125em] forced-colors:shr-fill-[CanvasText]"
                fill="currentColor"
                focusable="false"
                height="1em"
                role="img"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 448 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M224 256A128 128 0 1 0 224 0a128 128 0 1 0 0 256zm-45.7 48C79.8 304 0 383.8 0 482.3C0 498.7 13.3 512 29.7 512H418.3c16.4 0 29.7-13.3 29.7-29.7C448 383.8 368.2 304 269.7 304H178.3z"
                />
              </svg>
              <span>
                user-2
              </span>
            </span>
          </a>
          <span
            class="text-sm text-gray-600"
          >
            NaN-NaN-NaN NaN:NaN
          </span>
          <span
            class="text-sm flex items-center gap-1"
          >
            <svg
              aria-hidden="true"
              class="smarthr-ui-Icon group-[]/iconWrapper:shr-shrink-0 group-[]/iconWrapper:shr-translate-y-[0.125em] forced-colors:shr-fill-[CanvasText] text-yellow-500"
              fill="currentColor"
              focusable="false"
              height="1em"
              role="img"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 576 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M316.9 18C311.6 7 300.4 0 288.1 0s-23.4 7-28.8 18L195 150.3 51.4 171.5c-12 1.8-22 10.2-25.7 21.7s-.7 24.2 7.9 32.7L137.8 329 113.2 474.7c-2 12 3 24.2 12.9 31.3s23 8 33.8 2.3l128.3-68.5 128.3 68.5c10.8 5.7 23.9 4.9 33.8-2.3s14.9-19.3 12.9-31.3L438.5 329 542.7 225.9c8.6-8.5 11.7-21.2 7.9-32.7s-13.7-19.9-25.7-21.7L381.2 150.3 316.9 18z"
              />
            </svg>
            <span />
          </span>
        </div>
        <button
          class="disabled:shr-cursor-not-allowed disabled:shr-bg-clip-padding [&_.smarthr-ui-Icon]:forced-colors:disabled:shr-fill-[GrayText] shr-box-border shr-cursor-pointer shr-inline-flex shr-justify-center shr-items-center shr-gap-0.5 shr-text-center shr-whitespace-nowrap shr-rounded-m shr-border-shorthand shr-font-inherit shr-font-bold shr-leading-none focus-visible:shr-focus-indicator contrast-more:shr-border-high-contrast [&_.smarthr-ui-Icon]:shr-block [&_svg]:shr-block shr-p-0.5 shr-text-sm shr-min-h-[calc(theme(fontSize.sm)+theme(spacing.1)+theme(borderWidth.2))] shr-border-main shr-bg-main shr-text-white focus-visible:shr-border-main-darken focus-visible:shr-bg-main-darken hover:shr-border-main-darken hover:shr-bg-main-darken disabled:shr-border-main/50 disabled:shr-bg-main/50 disabled:shr-text-white/50 smarthr-ui-Button"
          type="button"
        >
          <span
            class="shr-min-w-0 shr-leading-[0]"
          >
            <div
              class="flex items-center gap-2"
            >
              <svg
                aria-hidden="true"
                class="smarthr-ui-Icon group-[]/iconWrapper:shr-shrink-0 group-[]/iconWrapper:shr-translate-y-[0.125em] forced-colors:shr-fill-[CanvasText]"
                fill="currentColor"
                focusable="false"
                height="1em"
                role="img"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 576 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M316.9 18C311.6 7 300.4 0 288.1 0s-23.4 7-28.8 18L195 150.3 51.4 171.5c-12 1.8-22 10.2-25.7 21.7s-.7 24.2 7.9 32.7L137.8 329 113.2 474.7c-2 12 3 24.2 12.9 31.3s23 8 33.8 2.3l128.3-68.5 128.3 68.5c10.8 5.7 23.9 4.9 33.8-2.3s14.9-19.3 12.9-31.3L438.5 329 542.7 225.9c8.6-8.5 11.7-21.2 7.9-32.7s-13.7-19.9-25.7-21.7L381.2 150.3 316.9 18z"
                />
              </svg>
              <span>
                いいね
              </span>
            </div>
          </span>
        </button>
      </div>
      <p>
        Second post
      </p>
    </li>
  </ul>
</DocumentFragment>
`;
